### PR TITLE
Allow for absolute finite difference step

### PR DESCRIPTION
**Pull request by divenex**
Allow for absolute, rather than only relative, step in the finite difference approximation of the derivatives. 

This is useful to include an option for **absolute** finite-difference steps in scipy/optimize/_lsq /least_squares.py

### CODE REVIEW

**Comment by divenex** [comment]
The user case that motivates this pull request is in the rather common situation when optimizing a function using logarithmic coordinates. In this case the use of relative finite-difference steps make no sense and absolute steps are required.

**Comment by ev-br** [comment]
cc @nmayorov

File: scipy/optimize/_lsq/least_squares.py [hunk 1]
```diff
@@ -868,8 +868,11 @@ def jac_wrapped(x, f):
                 tr_solver = 'lsmr'
 
     if method == 'lm':
-        result = call_minpack(fun_wrapped, x0, jac_wrapped, ftol, xtol, gtol,
-                              max_nfev, x_scale, diff_step)
+        if abs_step is None:
+            result = call_minpack(fun_wrapped, x0, jac_wrapped, ftol, xtol, gtol,
+                                  max_nfev, x_scale, rel_step)
+        else:
+            raise ValueError("'abs_step' can't be used when method='lm'")
```
Comment (nmayorov) [inline scipy/optimize/_lsq/least_squares.py]: Better move this exception above, around line 726 would be good. Another minor thing: end the sentence with a period (for consistency with other error messages).

**Comment by nmayorov** [comment]
@divenex good to see that you are using `least_squares`. Adding the absoulte step is reasonable, and your approach of introducing `abs_step` as another parameter looks simple and clear.

You definitely need to add/modify docstrings of `rel_step` and `abs_step` in the both functions. Explain what happens when the both parameters are None as well as when both are not None.

Also I'm afraid you have to extend / modify tests for `approx_derivative` and `least_squares`, it should be straightforward to do. If possible, account that `least_squares` tests are already taking a bit too long #5565 (by all means not your fault). Or would be good if you can clean up and speed up those tests in another PR.

Numerical differentiation was discussed for a long time with no definitive conclusion / solution. Here are two links:

1 https://github.com/scipy/scipy/pull/4884 --- discussion in PR with `approx_derivative`. 
2. https://gist.github.com/ev-br/7a1edb3f250bd375c46a --- Evgeni's overview / proposal.

So it's possible that this PR will be stalled / rejected in favor of a more general solution. I hope it won't discourage you from finishing it now. Maybe this PR will accelerate the discussion related to numerical differentiation, which would be very helpful by itself.

**Comment by nmayorov** [comment]
I have a question about what you wrote regarding optimization in log-scale variables. Say original variables span 10-10000 then log variables span 1-4. When using a relative step a final step in this range won't vary too strongly, so I don't see it is necessary going to hurt accuracy. Maybe you have a specific problem where a relative step failed?

I agree that a relative step is not a remedy (just a heuristics) and an absolute step can be just as useful. But perhaps a more general and overall better approach can be found.

**Comment by divenex** [comment]
In my applications I fit least squares functions that are not analytic nor computed to machine precision. They are the results of computations, where discretization and numerical "noise" is important.

For this reason I cannot rely on the standard automatic estimation of the step based on the power of EPS, but I have to specify all finite differencing steps by hand, based on my knowledge of the range of parameters over which I can trust any variation of the chi2 of my function. In practice, I optimize for the step size that gives the fastest convergence on a set of test cases. 

You are right that I can estimate the relative step that approximately matches my desired absolute step, within my range. And I may live with the variation of the step size of a factor of no more than an order of magnitude, within the logarithmic ranges of my function

But it just seemed cleaner to have the option to simply input an absolute step directly.

File: scipy/optimize/_lsq/least_squares.py [hunk 2]
```diff
@@ -726,6 +726,9 @@ def least_squares(
     if method == 'lm' and loss != 'linear':
         raise ValueError("method='lm' supports only 'linear' loss function.")
 
+    if method == 'lm' and abs_step is None:
```
Comment (nmayorov) [inline scipy/optimize/_lsq/least_squares.py]: not None (this is causing tests to fail).

Another nitpick: back ticks are used to reference parameters in error messages (not followed everywhere, but at least in this file). Note that this is how parameters should be referenced in docstrings.

**Comment by nmayorov** [comment]
I remember hesitating whether to use a relative or absolute step, probably decided to follow matlab' choice.

I agree that sometimes directly providing an absolute step is easier and makes more sense, and I personally don't mind including it to API. So I suggest to continue with this PR, other people can share their opinions later.

**Comment by divenex** [comment]
Thank you very much for the feedback @nmayorov. And let me take the opportunity to thank you for the work you did on least_squares. Congratulations! I have been totally impressed by the quality of your excellent and non-trivial effort.

**Comment by divenex** [comment]
Another common user case where absolute steps are preferred, is when one of the parameters we want to determine from our least squares fit is the position of an event (e.g. particle) in a detector. 

In this case the chi^2 varies by a similar amount when we change the coordinates by a fixed amount in detector pixels, regardless of the location on the detector. Using relative steps in this case can produce unnecessary orders of magnitude variations in the diff_step, eve tough the step should not depend on the particle position.

**Comment by nmayorov** [comment]
@divenex, thanks, the biggest reward for me is that people use it and (hopefully) get useful results.

Are you going to continue with tests and docstrings? A few tests for `abs_step` usage are required.

File: scipy/optimize/_lsq/least_squares.py [hunk 3]
```diff
@@ -233,7 +233,7 @@ def loss_function(f, cost_only=False):
 def least_squares(
         fun, x0, jac='2-point', bounds=(-np.inf, np.inf), method='trf',
         ftol=1e-8, xtol=1e-8, gtol=1e-8, x_scale=1.0, loss='linear',
-        f_scale=1.0, diff_step=None, tr_solver=None, tr_options={},
+        f_scale=1.0, abs_step=None, rel_step=None, tr_solver=None, tr_options={},
```
Comment (ev-br) [inline scipy/optimize/_lsq/least_squares.py]: Strictly speaking, renaming an argument name is a backcompat break :-(.

File: scipy/optimize/_numdiff.py [hunk 4]
```diff
@@ -334,7 +334,12 @@ def fun_wrapped(x):
     if np.any((x0 < lb) | (x0 > ub)):
         raise ValueError("`x0` violates bound constraints.")
 
-    h = _compute_absolute_step(rel_step, x0, method)
+    if abs_step is None:
+        h = _compute_absolute_step(rel_step, x0, method)
+    else:
+        # sign_x0 != 0 unlike np.sgn()
+        sign_x0 = (x0 >= 0).astype(float) * 2 - 1   
+        h = abs_step*sign_x0
```
Comment (ev-br) [inline scipy/optimize/_numdiff.py]: `np.where(x0 < 0, -1.0, 1.0)`

Comment (ev-br) [inline scipy/optimize/_numdiff.py]: So this ignores `rel_step` if `abs_step` is not None, correct? Better check and raise an exception (silently swallowing incorrect input is not nice).

**Comment by ev-br** [comment]
This now needs a rebase. I second @nmayorov's suggestions and defer to him to decide when it's ready.

**Pull request closed without merging**